go/pkg/client/exec.go

```diff
@@ -263,18 +256,10 @@ func (c *Client) ExecuteAndWaitProgress(ctx context.Context, req *repb.ExecuteRe
 				}
 			}
 		}
-		st := OperationStatus(lastOp)
-		if st != nil {
-			opError = true
-			if st.Code() == codes.DeadlineExceeded {
-				return nil
-			}
-			return st.Err()
-		}
 		return nil
 	}
 	err = c.Retrier.Do(ctx, func() error { return c.CallWithTimeout(ctx, "Execute", closure) })
-	if err != nil && !opError {
+	if err != nil {
```

Comment:
If I got this right, the intent of the changes is to not retry if operation if it returned the error which makes sense to me; however, doesn't this line introduce additional change in the behavior? 

Before the change, if `Retrier.Do` returned the error, but `opError` is false `ExecuteAndWait` would still return lastOp (the failed operation) . After the change `lastOp` is always discarded whenever err != nil, potentially removing part of the information from the client. Is this also an intended change?

Reply:
I don't think that's correct. Before this change, if Retrier.Do returned an error, *and* opError was false, we would take the branch and return (nil, err). It was only when opError was *true* that we'd skip the branch and return (lastOp, nil).

With the current change, if we get to the point where we previously would have checked for an operation failure, we know that Retrier.Do will return nil, so we won't take this branch.

(I find this code really, really hard to reason about...)

Reply:
You're right. Thank you for the explanation. Indeed, my concern doesn't seem to be valid.

Reply:
I agree this is terribly hard to read :)